### ISSUE TITLE
Intercom DOM elements inserted multiple times with Turbolinks 5

### ISSUE THREAD

**Issue by summera** [intercom/intercom-rails#268]
## Version info
  - intercom-rails version: `0.2.29`
  - Rails version: `4.2.8`

## Expected behavior
`#intercom-container` and `#intercom-frame` node elements are inserted into DOM once and only once.

## Actual behavior
Using Turbolinks 5, `#intercom-container` and `#intercom-frame` are being inserted into DOM multiple times during navigation. For example, when I start with a fresh page load, then click on a link, then click the back button in the browser, I am left with three `#intercom-frame` nodes and two `#intercom-container` nodes. Is the method to insert these nodes [idempotent](https://github.com/turbolinks/turbolinks#making-transformations-idempotent)?

Thanks for the help!

## Steps to reproduce
 1. Hard reload page
 2. Navigate by clicking a link
 3. Press back button in browser

## Logs
N/A

**Comment by aleksander-tatskiy** [comment]
I've the same issue like guys above (Turbolinks 5)

**Comment by summera** [comment]
@Skaelv wanted to check back in on this. Is this something you have seen before?

**Comment by meneerprins** [comment]
Also the same issue, any workarounds? (turbolinks 5, rails 5.0.6)

**Comment by tjdo** [comment]
I'm experiencing the same issue, looking into this as it breaks Intercom when navigating through screens.

**Comment by tjdo** [comment]
Just FYI, I contacted Intercom on this issue and they are now aware of and able to replicate this issue.

From Intercom:
"Looking into the issue we did manage to replicate this on our end using the Rails gem. We didn't test with manual code embed though (just the default intercom-rails installation)...

...That said, either way it would still be a bug we should address as soon as possible. I'll add the details you just shared with me to the issue and I'll let you know as soon as there is an update"

**Comment by summera** [comment]
@tjdo thanks!!

**Comment by summera** [comment]
@tjdo any update on this?

**Comment by tjdo** [comment]
@summera funny you mention it, I contacted them earlier today and there's no fix currently. I will be working with one of the engineers tomorrow to see if we can get a temporary fix, at least, since it has such a big impact on our integration.

**Comment by summera** [comment]
@tjdo Cool, let me know how that goes. We are reaching out to intercom as well to see if we can speed things up since it's been radio silence in this issue.

**Comment by tjdo** [comment]
Update: still working with Intercom on this issue. They are looking at our integration methods, and have not confirmed any ETA on a fix on their end. From all appearances, this is not a priority on their end as of yet.

**Comment by federicomoretti** [comment]
+1 Experiencing same problem.

**Comment by tjdo** [comment]
Please contact Intercom about the issue. I have a feeling the only way they will address this is if we put pressure on them. This is not an issue with intercom-rails, it's a problem in the Intercom library, which is closed source

**Comment by jgillman** [comment]
Contacted Intercom as well, currently waiting to hear back from their Support Engineers 👍

**Comment by summera** [comment]
FYI, we contacted intercom last week and they said they are tracking "internally". Not sure why we aren't getting updates in this issue though?

**Comment by choran** [comment]
Hi all, 
Apologies for not providing more updates here. As people have noted we are aware of the issue and thanks for the updates on it. We are looking at it internally but we have not been updating the issue here. So I will ensure that we provide updates on the public issue here as well as we work through this issue.
Cheers
Cathal

**Comment by meneerprins** [comment]
Thanks for the update, hope it can be addressed soon :)

**Comment by AleksandrLeontev** [comment]
@choran Hi. Any news? My client has uncomfortable cases with this bug on the project

**Comment by mrhead** [comment]
This is a long term issue. I've reported it on March 22nd, 2017 via Intercom widget itself. Let's hope it will be fixed soon.

**Comment by choran** [comment]
@AleksandrLeontev unfortunately, no ETA as of yet. It was a little trickier than we thought initially. Hopefully I will have more updates in the next few days. Either way I will let you know the progress.

**Comment by skylerwshaw** [comment]
Hi all 👋 

Apologies for the delay here. This was introduced when we resolved a separate issue on September 7 related to the Intercom messenger being removed from the DOM and not replaced in all cases. Unfortunately it resulted in this issue being opened 5 days later, where in some cases the JS would overzealously mount multiple instances into the DOM in response to Turbolinks changes.

The fix is in production and does not require an update to the `intercom-rails` gem as the issue was resolved in the Intercom JS.

**Comment by tjdo** [comment]
Thanks @skylerwshaw. I have tested this morning in our environment and our issues have been resolved.

**Comment by mrhead** [comment]
This still doesn't look fixed to me.

Steps to reproduce:

1) Visit a site with Intercom widget and Turbolinks.
2) Click on a link on the site.
3) Click on the back button.
4) Click on the forward button.
5) Repeat steps 3 and 4 for a few times.

Result: Multiple instances of Intercom widget exist on the page:

![intercom-widget-issue](https://user-images.githubusercontent.com/522375/33708775-ad2d8748-db3b-11e7-9178-09284f37c77a.gif)

**Comment by tjdo** [comment]
@mrhead I was unable to duplicate that behavior in my test environment. I'd be sure to turn off browser caching in dev tools, if not already.

**Comment by tjdo** [comment]
Setting for disabling cache when dev tools is open:
![image](https://user-images.githubusercontent.com/22504552/33731172-1e499da4-db37-11e7-9045-bb43cd5f89ed.png)

**Comment by summera** [comment]
> I'd be sure to turn off browser caching in dev tools, if not already.

@tjdo Maybe I'm missing something but why should browser caching have to be turned off for this to work correctly? Disabling caching kind of **defeats much of the purpose of Turbolinks**. This is why I had mentioned making the inserting of Intercom nodes [idempotent]([https://github.com/turbolinks/turbolinks#making-transformations-idempotent):

> ## Making Transformations Idempotent
> 
> Often you’ll want to perform client-side transformations to HTML received from the server. For example, you might want to use the browser’s knowledge of the user’s current time zone to group a collection of elements by date.
> 
> Suppose you have annotated a set of elements with `data-timestamp` attributes indicating the elements’ creation times in UTC. You have a JavaScript function that queries the document for all such elements, converts the timestamps to local time, and inserts date headers before each element that occurs on a new day.
> 
> Consider what happens if you’ve configured this function to run on `turbolinks:load`. When you navigate to the page, your function inserts date headers. Navigate away, and Turbolinks saves a copy of the transformed page to its cache. Now press the Back button—Turbolinks restores the page, fires `turbolinks:load` again, and your function inserts a second set of date headers.
> 
> To avoid this problem, make your transformation function _idempotent_. An idempotent transformation is safe to apply multiple times without changing the result beyond its initial application.
> 
> One technique for making a transformation idempotent is to keep track of whether you’ve already performed it by setting a `data` attribute on each processed element. When Turbolinks restores your page from cache, these attributes will still be present. Detect these attributes in your transformation function to determine which elements have already been processed.
> 
> A more robust technique is simply to detect the transformation itself. In the date grouping example above, that means checking for the presence of a date divider before inserting a new one. This approach gracefully handles newly inserted elements that weren’t processed by the original transformation.

Even if I disable caching though, I still get multiple `intercom-frame` nodes when navigating back and form with the back/forward browser buttons.
![screen shot 2017-12-07 at 11 43 05 am](https://user-images.githubusercontent.com/2474416/33732468-e2353abe-db43-11e7-989b-4591b071374d.png)

**Comment by tjdo** [comment]
The suggestion to turn off caching was simply to force the new version of the Intercom library to be downloaded, which comes from Intercom's CDN. Sorry for the confusion.

**Comment by tjdo** [comment]
@summera I agree with your approach. Our app does not use TL caching, we disabled it some time ago due to various issues we had with the caching. Which makes me wonder if the issue you guys are seeing is related to TL caching, because I am unable to duplicate it.

**Comment by skylerwshaw** [comment]
Digging into the repro instructions above. Looks like we may still have some issues with restoration visits. Working on a branch that approaches the issue in a different way. TL is notoriously difficult when managing iframes across transitions/caches/restorations. Will share some references when I get back to a desktop. The fix yesterday still takes care of route changes so many reported issues are solved, but the restoration activities surrounding the back and forward buttons are still somewhat affected. -Skyler

**Comment by summera** [comment]
@skylerwshaw any updates? This should still be an **open** issue right?

**Comment by skylerwshaw** [comment]
@mrhead @summera I've pushed additional changes to avoid having our messenger components cached. Can you confirm whether this is fixed for you in production?

**Comment by summera** [comment]
@skylerwshaw looks to be all good on my end. Thanks for you work on this!
